client.go

```diff
@@ -19,27 +19,12 @@ var ErrNoProtocols = errors.New("no protocols specified")
 // on this ReadWriteCloser. It returns an error if, for example,
 // the muxer does not know how to handle this protocol.
 func SelectProtoOrFail(proto string, rwc io.ReadWriteCloser) error {
-	errCh := make(chan error, 1)
-	go func() {
-		var buf bytes.Buffer
-		delimWrite(&buf, []byte(ProtocolID))
-		delimWrite(&buf, []byte(proto))
-		_, err := io.Copy(rwc, &buf)
-		errCh <- err
-	}()
-	// We have to read *both* errors.
-	err1 := readMultistreamHeader(rwc)
-	err2 := readProto(proto, rwc)
-	if werr := <-errCh; werr != nil {
-		return werr
-	}
-	if err1 != nil {
-		return err1
-	}
-	if err2 != nil {
-		return err2
+	if clearFn, err := setDeadline(rwc); err != nil {
```

Comment:
So, many connections _implement_ `net.Conn` (and `SetDeadline`) but don't _actually_ implement it (and instead return an error). This is why we do things like https://github.com/libp2p/go-libp2p-transport-upgrader/blob/0d4065ec7151149beef41f3a8b1fbf6978ed12dc/upgrader.go#L111-L132.

Reply:
Note: we could detect and handle `os.ErrNoDeadline`.